linode/obj/resource.go

```diff
@@ -104,12 +134,42 @@ func updateResource(ctx context.Context, d *schema.ResourceData, meta any) diag.
 		return putObject(ctx, d, meta)
 	}
 
+	cluster := d.Get("cluster").(string)
 	bucket := d.Get("bucket").(string)
 	key := d.Get("key").(string)
 	acl := s3types.ObjectCannedACL(d.Get("acl").(string))
 
 	if d.HasChange("acl") {
-		s3client, err := helper.S3ConnectionFromData(ctx, d, meta)
+		config := meta.(*helper.ProviderMeta).Config
+		client := meta.(*helper.ProviderMeta).Client
+
+		var objKeys ObjectKeys
+		objKeys.AccessKey = d.Get("access_key").(string)
+		objKeys.SecretKey = d.Get("secret_key").(string)
+
+		if !CheckObjKeysConfiged(objKeys) {
+			// If object keys don't exist in the resource configuration, firstly look for the keys from provider configuration
+			if providerKeys, ok := GetObjKeysFromProvider(objKeys, config); ok {
+				objKeys = providerKeys
+			} else if config.ObjUseTempKeys {
+				// Implicitly create temporary object storage keys
+				keys, diag := CreateTempKeys(ctx, client, bucket, cluster, "read_write")
+				if diag != nil {
+					return diag
+				}
+
+				objKeys.AccessKey = keys.AccessKey
+				objKeys.SecretKey = keys.SecretKey
+
+				defer CleanUpTempKeys(ctx, client, keys.ID)
+			}
+		}
+
+		if !CheckObjKeysConfiged(objKeys) {
+			return diag.Errorf("access_key and secret_key are required to update linode_object_storage_object")
+		}
+
+		s3client, err := helper.S3ConnectionFromData(ctx, d, meta, objKeys.AccessKey, objKeys.SecretKey)
```

Comment:
Since this section of checks appears a few times perhaps it'd be better to put in a helper function like `S3ConnectionFromData`?

Reply:
Or we might put a helper function to get keys in the `obj/helper`
Here is a sample code I wrote. I didn't test it though.
```go
func GetObjKeys(
	ctx context.Context,
	d *schema.ResourceData,
	config *helper.Config,
	client linodego.Client,
	bucket, cluster string,
) (ObjectKeys, diag.Diagnostics, func()) {
	objKeys := ObjectKeys{
		AccessKey: d.Get("access_key").(string),
		SecretKey: d.Get("secret_key").(string),
	}

	var teardownObjTempKeys func() = nil

	if !CheckObjKeysConfiged(objKeys) {
		// If object keys don't exist in the resource configuration, firstly look for the keys from provider configuration
		if providerKeys, ok := GetObjKeysFromProvider(objKeys, config); ok {
			objKeys = providerKeys
		} else if config.ObjUseTempKeys {
			// Implicitly create temporary object storage keys
			keys, diag := CreateTempKeys(ctx, client, bucket, cluster, "read_write")
			if diag != nil {
				return objKeys, diag, nil
			}

			objKeys.AccessKey = keys.AccessKey
			objKeys.SecretKey = keys.SecretKey
			teardownObjTempKeys = func() {CleanUpTempKeys(ctx, client, keys.ID)}
		}
	}

	if !CheckObjKeysConfiged(objKeys) {
		return objKeys, diag.Errorf(
			"access_key and secret_key are required to update linode_object_storage_object",
		), nil
	}
	return objKeys, nil, teardownObjTempKeys
}
```

Reply:
Thank you both for the suggestion! Just made this into one helper function.